### PR TITLE
feat: legacy html generator

### PR DESCRIPTION
**Pull request by ovflowd**
## Description

This PR introduces the Legacy HTML generator, which aims to generate an almost 1:1 replica of the existing https://nodejs.org/api/ documentation.

This PR introduces the fully-fledged `legacy-HTML` generator, with all the related utilities and tools to make complete API docs in HTML.

### Diversions from the original version:

- Some of the outputted HTML in the "view source" might be different
- The styles got a bit updated due to CSS conflicts with Shiki (Syntax Highlighter)
- A new syntax highlighter is being used (same as the Node.js Website), meaning token colors will differ.
- Sidebar Navigation is generated on-demand instead of relying on `nodejs/node/doc/api/index.md`
  - This means some of the entries are a bit different, although I believe they are fine.
- Copy button for CodeBoxes on the top instead of the bottom.
- Stability Index position is slightly below now (below the History Table and Source Link)

### Things to be done:

None. We're feature complete!

### Future generators

- `json` generator, being able to generate `.json` files for API docs
- `react` generator, the redesign of API docs generator.

## Validation

Clone, run `npm install`, then run the command against Node.js API docs, for example:

```bash
npx api-docs-tooling -i "/path/to/node/doc/api/*.md" -t legacy-html -o /tmp/out
```

You can just open the desired files on your browser.

### Preview

This is deployed temporarily on Netlify (not auto deployed on commits, I'm manually deploying it): https://node-api-docs-legacy-preview.netlify.app/

## Related Issues

Closes https://github.com/nodejs/api-docs-tooling/issues/15

### CODE REVIEW

File: jsconfig.json
Comment (ovflowd) [inline jsconfig.json:1]: I believe I can simply delete this file and things should still work, but Ill check that later today.

File: eslint.config.mjs
Comment (ovflowd) [inline eslint.config.mjs:6]: Unfortunately, new ESLint does not support nested comfigs, so I need to set web configuration here even tho only the html generator uses web APIs

Comment (targos) [reply eslint.config.mjs:6]: Are you actually using a browser to run the generator's code? Otherwise this is wrong and I would ask why you need it?

In any case, new ESLint supports parts of configs scoped to certain directories (using the `files` option, with glob matching).

Comment (ovflowd) [reply eslint.config.mjs:6]: No, Im not. As explained before pieces of the legacy-html generator have web-targeted aasets, such as /assets/api.js

ESLint complains about all Browser-APIs there because lack of target.

Comment (ovflowd) [reply eslint.config.mjs:6]: I know v8 supports that, but does v9? Any example of it? I can try to Google it too

Comment (targos) [reply eslint.config.mjs:6]: https://eslint.org/docs/latest/use/configure/configuration-files#specifying-files-and-ignores

Later in this array, add:

```
{
  files: ['src/generators/legacy-html/assets/*.js'],
  languageOptions: { globals: { ...globals.browser } }
},
```

**Comment by ovflowd** [comment]
cc @nodejs/collaborators and @nodejs/documentation as this PR changes a few aspects on the generated HTML as mentioned above.

**Comment by targos** [comment]
> Dropped support of `all.html`

Can this be avoided?

**Comment by ovflowd** [comment]
> > Dropped support of `all.html`
> 
> Can this be avoided?

Is there any concrete reason why? Afaik, nothing officially gets referenced to all.html; We can still write an all.json tho;

**Comment by targos** [comment]
Well I'm using it a lot (to CTRL+F because we don't have a good docs search feature), so I expect it to be the case for other people. This is a disruptive change that I would not like to see before the docs redesign.

**Comment by ovflowd** [comment]
> Well I'm using it a lot (to CTRL+F because we don't have a good docs search feature), so I expect it to be the case for other people. This is a disruptive change that I would not like to see before the docs redesign.

Gotcha, yes, the docs redesign aims to solve this. To be honest, the https://nodejs.org search box already includes all the API data, and it is continuously updated. (Also, the search box API data indexing will be done on the nodejs.org repository, so technically, it is decoupled from the actual API docs, or at least so it will be for the initial implementation.

For local dev environments, I believe it can source the results locally, but I'm not sure if they will match since Orama (our search provider) has some extra processing on their cloud.

I understand this is a pain point, but I do feel hesitant on an all.html, but I believe I can make it happen.

**Comment by targos** [comment]
> To be honest, the https://nodejs.org/ search box already includes all the API data

It only works for the latest LTS version (AFAIK)

**Comment by ovflowd** [comment]
> > To be honest, the [nodejs.org](https://nodejs.org/) search box already includes all the API data
> 
> It only works for the latest LTS version (AFAIK)

Correct, it indexes the LTS version. That's a good point. @micheleriva, what are the alternatives here? Can we make it for API doc generation so the whole database can be built locally based on metadata we ingest? i.e., is it a static database stored in a static file and loaded by the Orama client during bootstrap (client side)

File: src/utils/generators.mjs
Comment (canerakdas) [inline src/utils/generators.mjs:37]:
```suggestion
 * @TODO: This should not be necessary, and indicates errors within the API docs
```

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by ovflowd** [comment]
@targos I've added a `legacy-html-all` generator which extends on the `legacy-html` generator for generating the API docs + `all.html`

**Comment by ovflowd** [comment]
cc @nodejs/web-infra, can we have a 2nd thorough review and verify that the pages match?

I still need to update the code that generates the CodeBoxes and the Stability overview. I'll do that later tomorrow or today.

**Review by anonrig — APPROVED** [review]

**Comment by ovflowd** [comment]
> > Correct, it indexes the LTS version. That's a good point. @micheleriva, what are the alternatives here? Can we make it for API doc generation so the whole database can be built locally based on metadata we ingest? i.e., is it a static database stored in a static file and loaded by the Orama client during bootstrap (client side)
> 
> @ovflowd sorry for my late reply. Yes, we can definitely do this. Let me just point out that the search results may slightly differ from Orama Cloud (more performant, more accurate results) and Orama OSS. I hope that's not a dealbreaker.

I think that's a fair trade-off, at least within directly the Node.js API docs pages. Are there docs out there on how we can use this?

**Review by flakey5 — APPROVED** [review]

**Comment by micheleriva** [comment]
> I think that's a fair trade-off, at least within directly the Node.js API docs pages. Are there docs out there on how we can use this?

We have our docs at https://docs.orama.com/open-source, and you would probably need to use that plugin: https://docs.orama.com/open-source/plugins/plugin-data-persistence.

We can give you support probably starting next week or so to write down a reference architecture/implementation 🙏

**Review by JakobJingleheimer — APPROVED** [review]
This PR is quite massive (would it have been possible/feasible to break this up—at least smaller chunks into a feature branch that was then wholesale merged?), so I did not review every line, but the gist looks good and I approve of the items cited in the summary :)

Thanks Claudio!

**Comment by ovflowd** [comment]
> This PR is quite massive (would it have been possible/feasible to break this up—at least smaller chunks into a feature branch that was then wholesale merged?),

I think I could move the bug fixes to a different branch, yes, and the unrelated changes. I'll do this when this branch is feature-complete. Although the PR is open, I believe it should be a draft PR.

**Comment by ovflowd** [comment]
> > I think that's a fair trade-off, at least within directly the Node.js API docs pages. Are there docs out there on how we can use this?
> 
> We have our docs at [docs.orama.com/open-source](https://docs.orama.com/open-source), and you would probably need to use that plugin: [docs.orama.com/open-source/plugins/plugin-data-persistence](https://docs.orama.com/open-source/plugins/plugin-data-persistence).
> 
> We can give you support probably starting next week or so to write down a reference architecture/implementation 🙏

Awesome! cc @canerakdas or @AugustinMauroy can any of you open an issue to keep track of the search box work for this repo? For all's worth, it should be another generator to generate the Orama local database.

The search component itself would be either the same from the Node.js website or on the react generator here in the future; Butb the database generation should be a generator.

**Comment by ovflowd** [comment]
I've rebased the branch and am going to keep working on the features.

**Comment by ovflowd** [comment]
@AugustinMauroy tests are failing, I assume because I changed the way how stability indexes are built. Can you look into it?

**Comment by ovflowd** [comment]
cc @nodejs/web-infra another review please 🙏 I changed the HTML creation to be mostly

**Comment by ovflowd** [comment]
FYI: After the next review, I once again will need to decouple changes. Apologies for the prolonged review, implementing this is time consuming, and I want to ensure things work as expected.

**Comment by ovflowd** [comment]
FYI @nodejs/web-infra I split the changes: https://github.com/nodejs/api-docs-tooling/pull/66, waiting for review to getting it merged back here.

**Comment by ovflowd** [comment]
cc @nodejs/web-infra please add another round of reviews 🙏

**Comment by ovflowd** [comment]
> * There are changes in the name on the left menu. For example;
>   ```diff
>   - Assertion testing
>   + Assert
>   - C/C++ addons with Node-API
>   + Node-API
>   - Child processes
>   + Child Process
>   ```
> * There are differences in the listing in the left menu; `hr` should be after "Usage and example."

Have you read the PR description? 😅 -- navigation generation is done automatically from now on, instead from reading the manually defined navigation file that exists on /doc/api :P, so yes, it will be different, and that's intentional.

**Comment by ovflowd** [comment]
> Jump links don't exist/work

What are jump links?

**Comment by ovflowd** [comment]
> * The sequential links on the `/errors` page do not seem to be generated correctly;
> 
> image
> This issue is likely related to the above, but there are also problems with the type references on the following pages:
> 
> * `/url`
>   ```
>   <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object) | [`<string>`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type" class="type">&lt;Object&gt;</a>
>   ```
> * `/os`
>   ```
>   <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Number_type) (in MHz" class="type">&lt;number&gt;</a> 
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   This issue also appears on other pages
> 
> Except for these minor issues, it looks content-wise 1-1 identical to the current pages ✨

Fixed! Thanks for noticing it!

**Comment by canerakdas** [comment]
> > Jump links don't exist/work
> 
> What are jump links?

Anchors with a unique id located to the right of the headings;

image

**Comment by ovflowd** [comment]
> > > Jump links don't exist/work
> > 
> > 
> > What are jump links?
> 
> Anchors with a unique id located to the right of the headings;
> 
> image

Oh they were working before 🤔

File: src/generators/legacy-html-all/index.mjs
Comment (bmuenzenmeyer) [inline src/generators/legacy-html-all/index.mjs:66]: typo
```suggestion
    // for generating the `tableOfContents` with the `tableOfContents.parseNavigationNode` parser
```

File: src/generators/legacy-html/index.mjs
Comment (bmuenzenmeyer) [inline src/generators/legacy-html/index.mjs:53]: typo

```suggestion
    const groupedModules = groupNodesByModule(input);
```

File: src/generators/legacy-html/index.mjs
Comment (bmuenzenmeyer) [inline src/generators/legacy-html/index.mjs:104]: usage of typo
```suggestion
      const nodes = groupedModules.get(head.api);
```

**Review by bmuenzenmeyer — APPROVED** [review]
made a functional pass on Windows - works as described

found a couple typos while reading and the IDE yelling at me

**Review by canerakdas — APPROVED** [review]
LGTM!

**Review by aymen94 — APPROVED** [review]

**Review by araujogui — APPROVED** [review]

**Comment by ovflowd** [comment]
cc @nodejs/collaborators @nodejs/build this PR is finally ready, and all the features from the current legacy API docs template are built-in here.

Please, if possible test/run locally and check if the templates, and API docs look in a way y'all are satisfied with :)

**Comment by ovflowd** [comment]
cc @nodejs/web-infra for a final code review 🙏

File: src/generators/legacy-html/assets/style.css
Comment (AugustinMauroy) [inline src/generators/legacy-html/assets/style.css:1]:
```suggestion
/* --------------------- CSS Variables ---------------------------- */
```

Good practice is not to explain elements of a programming language. In this case, you shouldn't say it's a CSS variable.

**Review by AugustinMauroy — COMMENTED** [review]
LGTM !

File: src/generators/legacy-html/assets/style.css
Comment (ovflowd) [reply src/generators/legacy-html/assets/style.css:1]: You are aware I only copied the existing assets from node/node right?

Comment (ovflowd) [reply src/generators/legacy-html/assets/style.css:1]: Like, what's your point here? 😅

Comment (AugustinMauroy) [reply src/generators/legacy-html/assets/style.css:1]: Yeah, so it's the moment to update it

Comment (ovflowd) [reply src/generators/legacy-html/assets/style.css:1]: No, Augustin, we won't be touching or spending time on the assets of the legacy HTML pages; The purpose here is (of this PR) is to bring the old legacy api docs into the new api docs tooling into a functional state.

Not make nitpick adjustments on these old assets that are not going to be updated as we will start working on the new stuff.

Comment (ovflowd) [reply src/generators/legacy-html/assets/style.css:1]: Also this is the most, genuinely speaking, irrelevant nitpick that you could have provided here 😅

**Review by flakey5 — APPROVED** [review]

**Comment by targos** [comment]
More than the code, I'd like to review the output. Is there a preview somewhere or do I have to build it locally?

**Comment by ovflowd** [comment]
Unfortunately need to build locally. I'm not sure where I could upload a preview here. I recall you had a Netlify account or something? If you build you can simply upload the out/ folder

**Comment by ovflowd** [comment]
(Im not sure if I can deploy static files to Vercel)

**Comment by cclauss** [comment]
Do we have a good __input__ file that we can test with?

**Comment by ovflowd** [comment]
> Do we have a good **input** file that we can test with?

If you want to generate against all docs:

```bash
node bin/cli.mjs -i "/path/to/node/doc/api/*.md" -o /tmp/out -t legacy-html-all
```

**Comment by ovflowd** [comment]
@targos here's a deploy using Node v22.7.0 as a binary and targeting [this](https://github.com/nodejs/node/commit/7fea0108d57224c9aafbdf7c72d31f4b9da95c43) commit from `nodejs/node` for the doc files: https://66cb17954600d7f933d95110--candid-sprinkles-58d2ce.netlify.app/

**Comment by targos** [comment]
Little inconsistency with the index page's title:

image

**Comment by targos** [comment]
View on a single page option is missing (screenshot is from current docs), although the `all.html` page exists:

image

**Comment by targos** [comment]
> Every method has a "History" table always

I don't like that, because it seems that the history falls back to "Introduced in: v0.10.0", which is probably wrong in most cases.

**Comment by ovflowd** [comment]
> > Every method has a "History" table always
> 
> I don't like that, because it seems that the history falls back to "Introduced in: v0.10.0", which is probably wrong in most cases.

Could you give examples?

**Comment by ovflowd** [comment]
FYI @targos I've made a redeploy with a hot fix regarding the CJS/MJS pre elements.

**Comment by targos** [comment]
Is it possible to have a light theme for code blocks when the page is in light mode?

Current:
image


New:
image

**Comment by ovflowd** [comment]
> Little inconsistency with the index page's title (above is the new version):
> 
> image

Fixed

> View on a single page option is missing (screenshot is from current docs), although the `all.html` page exists:
> 
> image

Fixed

**Comment by ovflowd** [comment]
> Is it possible to have a light theme for code blocks when the page is in light mode?
> 
> Current: image
> 
> New: image

I could check on Shiki docs how they do that and I'll circle back to you :)

**Comment by ovflowd** [comment]
Fixes deployed on https://66cb2105b3f6bc035a2f6e2f--candid-sprinkles-58d2ce.netlify.app/

**Comment by targos** [comment]
> Sidebar Navigation is generated on-demand instead of relying on nodejs/node/doc/api/index.md
This means some of the entries are a bit different, although I believe they are fine.

I think that's fine (IIUC we can update the page headings to change the navigation text), except:
Now "Usage and example" is sorted with the other API docs pages, which doesn't make sense. I guess we're also lucky that "About" is sorted first ^^

**Comment by targos** [comment]
Consequence of the light mode issue:
image

**Comment by ovflowd** [comment]
> > Sidebar Navigation is generated on-demand instead of relying on nodejs/node/doc/api/index.md
> > This means some of the entries are a bit different, although I believe they are fine.
> 
> I think that's fine (IIUC we can update the page headings to change the navigation text), except: Now "Usage and example" is sorted with the other API docs pages, which doesn't make sense. I guess we're also lucky that "About" is sorted first ^^

Correct. Although, the reasoning behind this change is to not depend on the existence of an index.md to source from -- this tooling tries to be agnostic as much as possible.

In the future, with the new generator (redesign of API docs) the sidebar will be grouped by module type (Module, Class, Globals, etc) as it was for a short while on https://nodejs.dev

**Comment by ovflowd** [comment]
> Now "Usage and example" is sorted with the other API docs pages, which doesn't make sense.

Agreed. Although the synopsis page never made sense to me, and even less now with the Node.js Website Learn section. I think that page could be removed IMO.

**Comment by targos** [comment]
> > > Every method has a "History" table always
> > 
> > 
> > I don't like that, because it seems that the history falls back to "Introduced in: v0.10.0", which is probably wrong in most cases.
> 
> Could you give examples?

Forget it, it's actually part of the markdown source code. I just happened to click on three pages that had the same version by chance.

**Comment by ovflowd** [comment]
> > Is it possible to have a light theme for code blocks when the page is in light mode?
> > Current: image
> > New: image
> 
> I could check on Shiki docs. I believe they have support for dark/light theme modes.

Fixed.

> Consequence of the light mode issue: image

Fixed.

---

Re-deployed on same URL with Codebox fixes. (Including some margin between CJS and Copy Button)

**Comment by targos** [comment]
On https://node-api-docs-legacy-preview.netlify.app/async_context, compared with https://nodejs.org/docs/latest/api/async_context.html, the "Other versions" dropdown doesn't show version 16.x.
This page was introduced in v16.4.0.

Now I understand why https://github.com/nodejs/api-docs-tooling/pull/55#issuecomment-2308818808 was bugging me. The first "History" entry is misleading. It is about the introduction of the page in the documentation, not the introduction of the API.

**Comment by ovflowd** [comment]
> On https://node-api-docs-legacy-preview.netlify.app/async_context, compared with https://nodejs.org/docs/latest/api/async_context.html, the "Other versions" dropdown doesn't show version 16.x.
> 
> This page was introduced in v16.4.0.
> 
> 
> 
> Now I understand why https://github.com/nodejs/api-docs-tooling/pull/55#issuecomment-2308818808 was bugging me. The first "History" entry is misleading. It is about the introduction of the page in the documentation, not the introduction of the API.

Hmm, no, it is just the semver comparison I'm doing, I should fix that. It should definitely have on the v16 list.

**Comment by ovflowd** [comment]
But yes, the introduced_in I believe is when it got introduced in Node, and added, when it got added to the docs?

**Comment by targos** [comment]
I think we should not show a history table for `introduced_in`. This is jsut metadata that was added so we can build the version selector.

Sometime, like in the case of `AsyncLocalStorage`, we create a new page and move around the documentation. The new page then has an `introduced_in` version that doesn't correspond to the version that any API was introduced in. It's just the first version in which the page exists.

**Comment by ovflowd** [comment]
> I think we should not show a history table for introduced_in. This is just metadata that was added so we can build the version selector.

Aha, noted. Making changes.

**Comment by ovflowd** [comment]
I've fixed the semver comparison -- but wonder if we should remove "Introduced In" from the "History", because I also added the following to be part of the History table:

image

This was done, so that the metadata stays in one place and tom simplify it.

**Comment by targos** [comment]
I may be wrong, but I think `introduced_in` is like `source_link`, a special metadata that has nothing to do with API updates like the other ones in your screenshot.
It is only used once per markdown file, near the top.

**Comment by ovflowd** [comment]
> I may be wrong, but I think `introduced_in` is like `source_link` and `type`, a special metadata that has nothing to do with API updates like the other ones in your screenshot. It is only used once per markdown file, near the top.

@targos Ive updated the code so that the metadata generation is like on the current API docs. I will update the tests of this PR later.

**Comment by ovflowd** [comment]
@nodejs/web-infra @canerakdas one last code review? 🙏 

Also, @targos let me know if you find any other inconsistency 🙏

**Review by flakey5 — APPROVED** [review]

**Comment by ovflowd** [comment]
FYI I'merging this for the time being, to unblock @flakey5; Happy to iterate on following PRs. @AugustinMauroy I would love if you could help with some extra tests now that the logic is full I believe you have more material to test! 😃 

If you want you can also use the experimental snapshot feature for snapshot tests of html output based on mock markdown sources that mock numerous different pieces of our metadata and content (code blocks, switchable code blocks, all sort of yaml metadata etc)

I think that working on said tests would be awesome for you to get familiar with the nitty gritty of our API docs stuff ;D

**Comment by ovflowd** [comment]
@targos let me know if perchance I also missed anything. @flakey5 is working on the json generator (legacy-json and legacy-json-all) to generate matching json output for the current api docs :)

**Pull request merged**
